### ISSUE TITLE
hardtanh_backward_kernel implementation differs from documentation

### ISSUE THREAD

**Issue by james-p-xu** [pytorch/pytorch#120258]
### 🐛 Describe the bug

In the [PyTorch docs for Hardtanh](https://pytorch.org/docs/stable/generated/torch.nn.Hardtanh.html), taking the gradient of the piecewise provided yields: 1 for min_val <= x <= max_val, 0 otherwise.

```python
import torch

t = torch.tensor([-1.0, 0, 1.0], dtype = torch.float32)
t.requires_grad = True
t_htnh = torch.nn.functional.hardtanh(t)
t_htnh.backward(torch.ones_like(t_htnh))

print(t.grad)

# yields tensor([0., 1., 0.]), but should be tensor([1., 1., 1.])
```

The issue seems to be in the aten/src/ATen/native/cpu/Activation.cpp:hardtanh_backward_kernel function at line 812.

The line currently reads:
```python
return (float(self_val) <= min_val || float(self_val) >= max_val) ? scalar_t(0) : grad_val;
```
However, it should be (LTE -> LT, GTE -> GT):
```python
return (float(self_val) < min_val || float(self_val) > max_val) ? scalar_t(0) : grad_val;
```

since the equal cases should produce a gradient of 1.

### Versions

torch==2.0.1

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by mikaylagawarecki** [comment]
This looks right to me since if x == max_value or x == min_value, x is the output

Marking as actionable to send a fix

**Comment by lezcano** [comment]
The function is not differentiable at those points, so the gradient is not well-defined. See https://pytorch.org/docs/stable/notes/autograd.html#gradients-for-non-differentiable-functions.
